Switch from web3.js to ethers.js

codersjj opened the pull request:
Update Code Examples from Using Web3.js to Ethers.js

Resolves: #15759

codersjj committed:
docs: switch from web3.js to ethers.js

Update introduction-to-smart-contracts.rst

Update Code Examples from Using Web3.js to Ethers.js

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

remove trailing whitespaces

Update introduction-to-smart-contracts.rst

Update Code Examples from Using Web3.js to Ethers.js

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

docs: switch from web3.js to ethers.js

Update Code Examples from Using Web3.js to Ethers.js

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: matheusaaguiar <95899911+matheusaaguiar@users.noreply.github.com>

remove trailing whitespaces

Update introduction-to-smart-contracts.rst

Update Code Examples from Using Web3.js to Ethers.js

Update docs/introduction-to-smart-contracts.rst

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

docs: switch from web3.js to ethers.js

matheusaaguiar commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -181,21 +181,19 @@ arguments ``from``, ``to`` and ``amount``, which makes it possible to track
 
 To listen for this event, you could use the following
-JavaScript code, which uses `web3.js <https://github.com/web3/web3.js/>`_ to create the ``Coin`` contract object,
+JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``coin`` contract object,
```
Although it was already phrased like this before, maybe we could do something regarding that the text says the code "uses ether.js to **create** the coin contract object ..." but there is no line actually creating any object.
Maybe we should just change to:
```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract ,
```

matheusaaguiar replied:
Alternatively, we could also add a "templated" line of code which creates a ``Coin`` contract object:

```javascript
const coin = new ethers.Contract(coinAddress, abi, provider);
```

matheusaaguiar replied:
Does that make sense, @r0qs ?

matheusaaguiar commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -181,21 +181,19 @@ arguments ``from``, ``to`` and ``amount``, which makes it possible to track
 
 To listen for this event, you could use the following
-JavaScript code, which uses `web3.js <https://github.com/web3/web3.js/>`_ to create the ``Coin`` contract object,
+JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``coin`` contract object,
```
```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``Coin`` contract object,
```
Here `Coin` refers to the contract name, not the variable name in the example.

codersjj replied:
> Here `Coin` refers to the contract name, not the variable name in the example.

Got it, thanks for clarifying!

matheusaaguiar commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -181,21 +181,19 @@ arguments ``from``, ``to`` and ``amount``, which makes it possible to track
+    coin.on('Sent', async (from, to, amount, event) => {
+      console.log(`Coin transfer: ${ethers.formatUnits(amount, 18)} coins were sent from ${from} to ${to}.`);
+    
+      const senderBalance = await contract.balances(from);
+      const receiverBalance = await contract.balances(to);
```
```suggestion
      const senderBalance = await coin.balances(from);
      const receiverBalance = await coin.balances(to);
```

I gues this should be ``coin`` as that is the name used before to listen to the event, right?

codersjj replied:
> I gues this should be `coin` as that is the name used before to listen to the event, right?

Good catch, thanks!

matheusaaguiar reviewed (commented):
Overall looks good, but I think we should address some questions/suggestions before.

matheusaaguiar reviewed (commented):
Also, [this is failing](https://app.circleci.com/pipelines/github/ethereum/solidity/37781/workflows/1f4556ed-647a-4ae2-a674-a52495c32b4c/jobs/1734392) because there are trailing whitespaces that needed to be removed in lines `191` and `194`

codersjj commented:
> Also, [this is failing](https://app.circleci.com/pipelines/github/ethereum/solidity/37781/workflows/1f4556ed-647a-4ae2-a674-a52495c32b4c/jobs/1734392) because there are trailing whitespaces that needed to be removed in lines `191` and `194`

Fixed

codersjj commented:
> @codersjj can you rebase and squash the commits, please?

That's alright. I've rebased and squashed the commits. However, I noticed an extra merge commit was generated. Is this acceptable, or should I clean it up? Let me know! 😊

cameel commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -181,21 +181,19 @@ arguments ``from``, ``to`` and ``amount``, which makes it possible to track
+      const senderBalance = await coin.balances(from);
+      const receiverBalance = await coin.balances(to);
+
+      console.log(`Balances now:\nSender: ${ethers.formatUnits(senderBalance, 18)}\nReceiver: ${ethers.formatUnits(receiverBalance, 18)}`);
```
The original formatting here was more readable. No reason to remove it.
```suggestion
        console.log(
            `Balances now:\n` +
            `Sender: ${ethers.formatUnits(senderBalance, 18)}\n` +
            `Receiver: ${ethers.formatUnits(receiverBalance, 18)}`
        );
```

cameel commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -181,21 +181,19 @@ arguments ``from``, ``to`` and ``amount``, which makes it possible to track
 
 To listen for this event, you could use the following
-JavaScript code, which uses `web3.js <https://github.com/web3/web3.js/>`_ to create the ``Coin`` contract object,
+JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract ,
```
```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract,
```

cameel commented on docs/introduction-to-smart-contracts.rst:
Note that this is not the only place where we're using web3.js. We have a few more examples/mentions:
- [Events](https://docs.soliditylang.org/en/develop/contracts.html#events)
- [Micropayment channel example](https://docs.soliditylang.org/en/develop/solidity-by-example.html#micropayment-channel)
- [Creating contracts](https://docs.soliditylang.org/en/develop/contracts.html#creating-contracts)

We should probably also mention deprecation in the "Resources" section.

matheusaaguiar replied:
I think there's also a couple of mentions/examples in [Solidity by Example](https://docs.soliditylang.org/en/v0.8.29/solidity-by-example.html) that could be adjusted.
@codersjj , are you still interested in working this?
Sorry for the long time to come back to this. Hopefully we can merge it soon now.

cameel reviewed (changes requested):
Thanks for the contribution! The code looks good overall.

However, the PR does not seem to be addressing #15759 completely (despite the `Resolves: #15759` in the description). There are a few more web3.js snippets in the docs and also some mentions of the library.

I don't think it really makes sense to convert just one example and leave everything else as is. Having one in some places and the other in others might be even worse than just leaving it as is.

cameel commented:
> Sure! I've rebased and squashed the commits. However, I noticed an extra merge commit was generated. Is this acceptable, or should I clean it up? Let me know! 😊

Rebasing would get rid of the merge commit. I think you must have merged `develop` on your branch instead of rebasing it on `develop`. Also, the commits with review fixes do not seem squashed.

codersjj commented:
> Thanks for the contribution! The code looks good overall.
> 
> However, the PR does not seem to be addressing #15759 completely (despite the `Resolves: #15759` in the description). There are a few more web3.js snippets in the docs and also some mentions of the library.
> 
> I don't think it really makes sense to convert just one example and leave everything else as is. Having one in some places and the other in others might be even worse than just leaving it as is.

Thanks for pointing that out—I completely agree! 🙌 You're right that consistency is key here. I'll go through the docs to find any remaining web3.js snippets/mentions and update them to use ethers.js instead. Appreciate the thorough review, and please feel free to flag any other instances I might miss!
